lib/block-supports/layout.php

```diff
@@ -412,16 +433,16 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
 		}
 	}
 
-	/*
-	 * This assumes the hook only applies to blocks with a single wrapper.
-	 * A limitation of this hook is that nested inner blocks wrappers are not yet supported.
-	 */
-	$content = preg_replace(
-		'/' . preg_quote( 'class="', '/' ) . '/',
-		'class="' . esc_attr( implode( ' ', $class_names ) ) . ' ',
-		$block_content,
-		1
-	);
+	$inner_content_classnames = gutenberg_identify_inner_block_wrapper_classnames( $block['innerContent'] );
+
+	$content = new WP_HTML_Tag_Processor( $block_content );
+	if ( $inner_content_classnames ) {
+		$content->next_tag( array( 'class_name' => $inner_content_classnames ) );
```

Comment:
this does what you want it to do, but it surprised me that it did, because we intended to search for individual class names and not the entire list of classes. it would be good to be aware that if we were to chop out one of the classes this would fail because the classes in `$inner_content_classnames` would no longer be an identical string of the same exact length as the `class` attribute in the searched HTML

Reply:
I think it's safe to take that risk here because we haven't changed anything in `$block_content` up until this point, and we don't have a better way of targeting the block's inner wrapper. Potentially, if we had a way to know ahead of time which tag is the last, we could add a classname to it instead of storing the existing ones, and then remove it after this step.

Reply:
@tellthemachines you could filter the tag openers manually:

```php
$sought_class_names = $inner_content_classnames;
while( $content->next_tag() ) {
    $tag_class_names = preg_split('/\s+/', $content->get_attribute( 'class' ));
    if ( /* the two overlap */ ) {
      // ...
    }
}
```

I dislike the `preg_split`, though. @dmsnell We already have `get_attribute()` in the processor, any reason not to add `get_classes()`?

Reply:
@adamziel my spidey-sense suggests it wouldn't be that much of an added value above `get_attribute()`. perhaps `has_class()` though would be.

what I _don't_ want people to do is what we have in your code example. we provide semantic operations to add and remove CSS classes. if we're encouraging people to extract the raw value(s) and perform their own twiddling then we're just pushing them away from a robust framework and towards multiple ad-hoc implementations of the same thing.

it highlights the _elephant in the room_, which are class names with HTML entities. we may have to come back and revisit those; maybe we could find a way to detect if we think we have an entity (a character is `&`) and then rewind the class comparison with a cloned and decoded value.

point being is that I think still it's more ideal to focus on intended behaviors instead of internal mechanics. _why_ do we want `get_classes()` when we can already query by tags containing a class name and `get_attribute( 'class' )`

Reply:
Such a good point @dmsnell, let's explore adding `has_class`. I added a note to the overview issue's description so we don't lose track of it.

Reply:
Here's a good example of where this would be useful

https://github.com/WordPress/gutenberg/pull/40778/commits/baef2619d79a1fcdd54c1b0d4dfdc64d481a4d76

I was assuming it already had such a method!

Reply:
+1 for an idiomatic `has_class` as the one-line implementation we're likely going to see more of does not account for nuances like `class="not-current-menu-item"`:

```php
if ( strpos( $inner_blocks_html, 'current-menu-item' ) ) {
     // ...
}
```

cc @dmsnell

Reply:
@dmsnell @aristath @getdave here's a PR that adds `has_class` https://github.com/WordPress/gutenberg/pull/46232

Reply:
@getdave what is the good example in that `bugfix` link? what was the bug? I'm trying to understand how `has_class()` would be a help there.

taking a guess, it looks like the goal is to only seek out `wp-block-navigation-item__content` if there is a `current-menu-item` in the document, and I'm not sure where `has_class()` comes in there.

if we want to know if any inner block HTML contains the class, the tag processor already includes this functionality.

```php
$inner_tags = new WP_HTML_Tag_Processor( $inner_blocks_html );
$has_current_menu_item = $inner_tags->next_tag( [ 'class_name' => 'current-menu-item' ] );
```

Secondly, I don't know that the `strpos()` check is all that bad. If we're looking to make sure that we don't introduce needless slow-down, then doing a quick string search over the inner HTML _for the possibility of finding the class name_ might be a good way to keep lots of requests fast while only resorting to proper HTML parsing when we have reason to believe the class might exist within.

Reply:
@dmsnell Using `next_tag` like that is very clever! And now that I think about it, it's also fully within what it was invented for, it just didn't jump out to me in the first place. 

> I don't know that the `strpos()` check is all that bad.

Now that I think about it, the `class_name` (and `tag_name`) matcher could perform that `strpos` before trying to match tags.

Reply:
@adamziel apart from the HTML entities, which we currently don't match, the `strpos` speedup might warrant more justification before putting it in the tag processor.

on the one hand, `strpos()` is going to be faster if the wanted class isn't in plaintext in the document, but on the other hand, it scans up to the end of the document every time, and if we expect to find another tag "soon" after where the current pointer is, there's no need to repeatedly iterate through the document.

in terms of the general case I'd really want measured data showing that the additional complexity is worth it. my guess is that sometimes it would be faster and sometimes it would be slower, but in all cases it would add additional complexity.

Reply:
https://github.com/WordPress/wordpress-develop/pull/5096